### PR TITLE
Fix ONNX Runtime inference in `ORTTrainer`

### LINKED ISSUE #692: Inference with ORT failed with Bloom-560m model
### System Info

```shell
Optimum - latest from github
Python - 3.8
platform NVidia V100
```

### Who can help?

@JingyaHuang 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

1. clone from Optimum under, say /scratch/workspaceblobstore/users/ettao/opt-bloom dir
2. cd optimum && python setup install
3. run this cmd:

python -m torch.distributed.launch --nproc_per_node=8 /ws/optimum/examples/onnxruntime/training/language-modeling/run_clm.py --model_name_or_path bigscience/bloom-560m --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --label_smoothing 0.1 --do_eval --output_dir /dev/shm --overwrite_output_dir --max_steps 200 --logging_steps 20 --per_device_eval_batch_size 2 --fp16 --inference_with_ort

Getting error:

[ERROR!] Evaluation with ONNX Runtime is not available for bigscience/bloom-560m model. Set `inference_with_ort=False` to evaluate with PyTorch.
Traceback (most recent call last):
  File "/scratch/workspaceblobstore/users/ettao/opt-bloom/optimum/examples/onnxruntime/training/language-modeling/run_clm.py", line 593, in <module>
    main()
  File "/scratch/workspaceblobstore/users/ettao/opt-bloom/optimum/examples/onnxruntime/training/language-modeling/run_clm.py", line 559, in main
    metrics = trainer.evaluate(inference_with_ort=inference_args.inference_with_ort)
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/optimum-1.6.2.dev0-py3.8.egg/optimum/onnxruntime/trainer.py", line 803, in evaluate
    output = eval_loop(
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/optimum-1.6.2.dev0-py3.8.egg/optimum/onnxruntime/trainer.py", line 974, in evaluation_loop_ort
    ort_model = ort_model_cls.from_pretrained(model_id=model_id, file_name=file_name)
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/optimum-1.6.2.dev0-py3.8.egg/optimum/onnxruntime/modeling_ort.py", line 557, in from_pretrained
    return super().from_pretrained(
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/optimum-1.6.2.dev0-py3.8.egg/optimum/modeling_base.py", line 325, in from_pretrained
    return from_pretrained_method(
TypeError: _from_pretrained() got an unexpected keyword argument 'file_name'
[Errno 2] No such file or directory: '/dev/shm/transformer.word_embeddings.weight'

### Expected behavior

ORT team would like to do inferencing benchmark for Bloom-560m between

torch baseline vs. ORT inferencing

and possibly identify areas to improve on ORT. It will be convenient if the cmd above would work for bloom-560m model (and a few more larger bloom models).
I hope the cmd is correct, and if there's any workaround for now, it will be great.
Thanks.

### PR DESCRIPTION
**Pull request by JingyaHuang**
# What does this PR do?

As suggested by the title, the inference with ONNX runtime is currently broken due to recent changes:
* Migration of exporter(from transformers to optimum).
* Using subclasses of `ORTModel` instead of bare inference session for inference.

Fixes #692 #707

This PR fixes the following:
* Missing config(now the config will also be saved even without training, while exporting ONNX model)
* Remove redundant codes for numpy->torch tensor, as we use `ORTModel` now.
* Adapt export model names to the default of `ORTModel`.
* Validate execution provider while using `onnxruntime-training` for inference.
* IOBinding with dlpack while using `onnxruntime-training`.

* And also update to be compatible with transformers 4.26.0

__[CAVEAT]__

Due to current limit of `ORTModelForCausalLM`(*doesn't support `loss` as an output*), the inference for decoding will not be able to leverage pre-computed attention keys and values, which means significant overhead. This will be fixed after a refactoring, so to benchmark the performance of ONNX runtime inference, please use subclasses of `ORTModel` instead of `ORTTrainer`.

### CODE REVIEW

File: optimum/onnxruntime/io_binding/io_binding_helper.py [hunk 1]
```diff
@@ -140,7 +140,7 @@ def to_pytorch_via_cupy(ort_value: OrtValue) -> torch.Tensor:
     def to_pytorch_via_dlpack(ort_value: OrtValue) -> torch.Tensor:
         from torch._C import _from_dlpack
 
-        torch_tensor = ort_value.to_dlpacks(_from_dlpack)
+        torch_tensor = _from_dlpack(ort_value.to_dlpack())
```
Comment (fxmarty) [inline optimum/onnxruntime/io_binding/io_binding_helper.py:143]: What is the difference?

File: optimum/onnxruntime/trainer.py [hunk 2]
```diff
@@ -946,17 +956,17 @@ def evaluation_loop_ort(
...
             logger.info("[INFO] Exporting the model to ONNX...")
             if self.args.deepspeed and self.args.fp16:
                 export_device = "cuda"
             else:
                 export_device = "cpu"
 
-            with_loss = has_labels and not self.label_smoother
+            with_loss = (
+                has_labels and not self.label_smoother
+            )  # With `label_smoother` the loss will be computed outside modeling
```
Comment (fxmarty) [inline optimum/onnxruntime/trainer.py:969]:
```suggestion
            # With `label_smoother` the loss will be computed outside modeling
            with_loss = has_labels and not self.label_smoother
```

File: optimum/onnxruntime/trainer.py [hunk 3]
```diff
@@ -1195,10 +1208,16 @@ def prediction_loop_ort(
         else:
             ort_model_cls = ORTModelForCustomTasks
 
-        model_id, file_name = os.path.split(self.onnx_model_path)
-        ort_model = ort_model_cls.from_pretrained(model_id=model_id, file_name=file_name)
-
+        model_id = self.onnx_model_path
         args = self.args
+        # Temporary fix for decoder, now `use_cache` set to False which
```
Comment (fxmarty) [inline optimum/onnxruntime/trainer.py:1213]: which?

File: optimum/onnxruntime/utils.py [hunk 4]
```diff
@@ -187,6 +187,11 @@ def validate_provider_availability(provider: str):
                     raise ImportError(
                         f"`onnxruntime-gpu` is installed, but GPU dependencies are not loaded. It is likely there is a conflicting install between `onnxruntime` and `onnxruntime-gpu`. Please install only `onnxruntime-gpu` in order to use {provider}."
                     )
+                elif os.path.isfile(path_cuda_lib) and is_onnxruntime_training_available():
+                    if provider == "TensorrtExecutionProvider":
+                        raise ImportError(
+                            f"Asked to use {provider}, but `onnxruntime-training` package doesn't support {provider}. Please use `CUDAExecutionProvider` instead."
+                        )
```
Comment (fxmarty) [inline optimum/onnxruntime/utils.py:194]: Are `onnxruntime-gpu` and `onnxruntime-training` incompatible? Somebody may have `onnxruntime-training` installed but still want to use TRT execution provider for inference.

File: optimum/onnxruntime/trainer.py [hunk 5]
```diff
@@ -1460,18 +1473,36 @@ def _export(
...
+        output_path = model_path / ONNX_DECODER_NAME if is_decoder else model_path / ONNX_WEIGHTS_NAME
+        _ = export(
+            model=model,
+            config=onnx_config,
+            opset=opset,
+            output=output_path,
+            device=device,
+        )
+
+        model.config.save_pretrained(model_path)
```
Comment (fxmarty) [inline optimum/onnxruntime/trainer.py:1505]: Out of curiosity, what happens for encoder/decoder models? Do we save as monolith?

File: optimum/onnxruntime/trainer.py [hunk 6]
```diff
@@ -1171,17 +1184,17 @@ def prediction_loop_ort(
...
             logger.info("[INFO] Exporting the model to ONNX...")
             if self.args.deepspeed and self.args.fp16:
                 export_device = "cuda"
             else:
                 export_device = "cpu"
 
-            with_loss = has_labels and not self.label_smoother
+            with_loss = (
+                has_labels and not self.label_smoother
+            )  # With `label_smoother` the loss will be computed outside modeling
```
Comment (fxmarty) [inline optimum/onnxruntime/trainer.py:1197]: same as previous comment (format)

**Comment by fxmarty** [comment]
Looks great! Could you maybe add a nightly test? Basically to make sure that what was not working is now tested so that we can quickly catch if something breaks.

File: optimum/onnxruntime/io_binding/io_binding_helper.py (hunk 1 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/io_binding/io_binding_helper.py:143]: No difference technically, just a change to adapt the current API in onnxruntime.

File: optimum/onnxruntime/utils.py (hunk 4 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/utils.py:194]: TRT ep is not available with `onnxruntime-training`.

File: optimum/onnxruntime/trainer.py (hunk 5 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/trainer.py:1504]: This is the fix for `ORTTrainer`, the fix for `ORTSeq2SeqTrainer` is planned in another PR.

File: optimum/onnxruntime/trainer.py (hunk 3 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/trainer.py:1212]: Missing part added.

File: optimum/onnxruntime/utils.py (hunk 4 above)
Comment (fxmarty) [reply optimum/onnxruntime/utils.py:194]: Are onnxruntime-gpu and onnxruntime-training incompatible?

File: optimum/onnxruntime/trainer.py (hunk 5 above)
Comment (fxmarty) [reply optimum/onnxruntime/trainer.py:1503]: I see thanks

File: optimum/onnxruntime/utils.py (hunk 4 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/utils.py:194]: Yes.

Comment (fxmarty) [reply optimum/onnxruntime/utils.py:194]: All good then

File: optimum/onnxruntime/trainer.py [hunk 7]
```diff
@@ -970,10 +979,16 @@ def evaluation_loop_ort(
...
             ort_model_cls = ORTModelForCustomTasks
 
-        model_id, file_name = os.path.split(self.onnx_model_path)
-        ort_model = ort_model_cls.from_pretrained(model_id=model_id, file_name=file_name)
-
+        model_id = self.onnx_model_path
         args = self.args
+        # Temporary fix for decoder, now `use_cache` set to False which
+        # TODO: Use cache once `ORTModelForCausalLM` supports `loss` as output
+        if ort_model_cls == ORTModelForCausalLM:
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/trainer.py:986]:
```suggestion
        if ort_model_cls is ORTModelForCausalLM:
```

File: optimum/onnxruntime/trainer.py [hunk 8]
```diff
@@ -1195,10 +1206,16 @@ def prediction_loop_ort(
...
             ort_model_cls = ORTModelForCustomTasks
 
-        model_id, file_name = os.path.split(self.onnx_model_path)
-        ort_model = ort_model_cls.from_pretrained(model_id=model_id, file_name=file_name)
-
+        model_id = self.onnx_model_path
         args = self.args
+        # Temporary fix for decoder, now `use_cache` set to False which cause overhead on re-computing past keys and values.
+        # TODO: Use cache once `ORTModelForCausalLM` supports `loss` as output
+        if ort_model_cls == ORTModelForCausalLM:
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/trainer.py:1213]:
```suggestion
        if ort_model_cls is ORTModelForCausalLM:
```

File: optimum/onnxruntime/trainer.py [hunk 9]
```diff
@@ -970,10 +979,16 @@ def evaluation_loop_ort(
...
+        model_id = self.onnx_model_path
         args = self.args
+        # Temporary fix for decoder, now `use_cache` set to False which
+        # TODO: Use cache once `ORTModelForCausalLM` supports `loss` as output
+        if ort_model_cls == ORTModelForCausalLM:
+            ort_model = ort_model_cls.from_pretrained(
+                model_id=model_id, use_cache=False, provider="CUDAExecutionProvider"
+            )
+        else:
+            ort_model = ort_model_cls.from_pretrained(model_id=model_id, provider="CUDAExecutionProvider")
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/trainer.py:991]:
```suggestion
       ort_model = ort_model_cls.from_pretrained(
                model_id=model_id, use_cache=ort_model_cls is not ORTModelForCausalLM, provider="CUDAExecutionProvider"
            )
```

**Review by michaelbenayoun — APPROVED** [review]
LGTM

**Comment by fxmarty** [comment]
Highlighting it:

> Could you maybe add a nightly test?

I really think it is important to test what was not tested, was failing while it should not. Or open an issue to track it so as to do it in a following PR.

File: optimum/onnxruntime/trainer.py (hunk 9 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/trainer.py]: Other subclasses of `ORTModel` don't have `use_cache` argument will fail.

**Comment by JingyaHuang** [comment]
@fxmarty For training tests, some of them are outdated, I would prefer to tackle it in another PR.

**Pull request merged**